swarm/swap_test.go

```diff
@@ -0,0 +1,423 @@
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		//wait for kademlia to be healthy
+		if _, err := sim.WaitTillHealthy(ctx, 2); err != nil {
+			return err
+		}
+
+		nodeIDs := sim.UpNodeIDs()
+		rand.Shuffle(len(nodeIDs), func(i, j int) {
+			nodeIDs[i], nodeIDs[j] = nodeIDs[j], nodeIDs[i]
+		})
+		//upload a file for every node
+		for _, id := range nodeIDs {
+			item, ok := sim.NodeItem(id, bucketKeySwarm)
+			if !ok {
+				log.Error("No swarm")
+				return errors.New("No swarm")
+			}
+			swarm := item.(*Swarm)
+			key, data, err := uploadFile(swarm)
+			if err != nil {
+				return err
+			}
+			log.Trace("file uploaded", "node", id, "key", key.String())
+			files = append(files, file{
+				addr:   key,
+				data:   data,
+				nodeID: id,
+			})
+		}
+
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		for {
+			if retrieve(sim, files, &checkStatusM, &nodeStatusM, &totalFoundCount) == 0 {
+				break
+			}
+		}
+
+		time.Sleep(5 * time.Second)
+		//every node has a map to all nodes it had interactions
+		//each entry in the map is a map of the other node with all the balances
+		balancesMap := make(map[enode.ID]map[enode.ID]int64)
+
+		//iterate all nodes
+		for _, node := range sim.NodeIDs() {
+			item, ok := sim.NodeItem(node, bucketKeySwarm)
+			if !ok {
+				log.Error("No swarm")
+				return errors.New("No swarm")
+			}
+			swarm := item.(*Swarm)
+
+			//submap for each node is a map of all nodes with the balance for that node
+			subBalances := make(map[enode.ID]int64)
+
+			//iterate all nodes again...
+			//get all balances with other peers for every node
+			for _, n := range sim.NodeIDs() {
+				if node == n {
+					continue
+				}
+
+				//get the peer's balance with this node
+				balance, err := swarm.swap.GetPeerBalance(n)
+				if err == nil {
+					subBalances[n] = balance
+					log.Debug(fmt.Sprintf("Balance of node %s to node %s: %d", node.TerminalString(), n.TerminalString(), balance))
+				} else {
+					log.Debug(fmt.Sprintf("Node %s has no balance with node %s", node.TerminalString(), n.TerminalString()))
+				}
+			}
+			//update the map for this node
+			balancesMap[node] = subBalances
+		}
+
+		//print all the balances if requested
+		if *printStats {
+			for k, v := range balancesMap {
+				fmt.Println(fmt.Sprintf("node %s balances:", k.TerminalString()))
+				for kk, vv := range v {
+					fmt.Println(fmt.Sprintf(".........with node %s: balance %d", kk.TerminalString(), vv))
+				}
+			}
+		}
+
+		//now iterate the whole map
+		//and check that every node k has the same
+		//balance with a peer as that peer with the node,
+		//but in inverted signs
+
+		//iterate the map
+		success := true
+		for k, mapForK := range balancesMap {
+			//iterate the submap
+			for n, balanceKwithN := range mapForK {
+				//iterate the main map again
+				for subK, mapForSubK := range balancesMap {
```

Comment:
no need to iterate if you care about `n==subK` case only

Reply:
I have edited the code to what I believe would be correct (thanks for this improvement suggestion), but please verify it is actually correct. Thank you.